crates/fs/public/directory.rs

```diff
@@ -0,0 +1,901 @@
+//! Public fs directory node.
+
+use std::{
+    cmp::Ordering,
+    collections::BTreeMap,
+    future::Future,
+    io::{Read, Seek},
+    rc::Rc,
+};
+
+use crate::{error, shared, BlockStore, FsError, Metadata, Shared, UnixFsNodeKind};
+use anyhow::{bail, ensure, Result};
+use async_recursion::async_recursion;
+use chrono::{DateTime, Utc};
+use field_names::FieldNames;
+use libipld::{
+    cbor::{cbor::MajorKind, decode, encode, DagCborCodec},
+    codec::{Decode, Encode},
+    Cid, IpldCodec,
+};
+
+use super::{Id, Link, PublicFile, PublicNode};
+
+//--------------------------------------------------------------------------------------------------
+// Type Definitions
+//--------------------------------------------------------------------------------------------------
+
+/// A directory in a WNFS public file system.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct PublicDirectory(Shared<PublicDirectoryInner>);
+
+#[derive(Debug, Clone, PartialEq, Eq, FieldNames)]
+struct PublicDirectoryInner {
+    metadata: Metadata,
+    userland: BTreeMap<String, Link>,
+    previous: Option<Cid>,
+}
+
+/// Represents a directory that has possibly diverged. It is the result of operating on a directory.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct OpResult<T> {
+    // The root node. It is the same as the previous root node if the directory has not been diverged.
+    pub root_node: Shared<PublicNode>,
+    // Implementation dependent but it usually the last leaf node operated on.
+    pub result: T,
+    /// Whether this is a divergence or not.
+    pub diverged: bool,
+}
```

Comment:
I think we might be able to go without the `OpResult` type.
* `get_node` returns an `OpResult`, but that means you get back a `PublicNode` in `root_node` and `result`, where I'd pretty much expect `root_node` to be the same thing that that's used to call `get_node` on.
* `read` returns an `OpResult`, but like `get_node`, it returns back the `root_node` you originally called `read` on.
* `write` returns the parent of the written node. That `result` isn't used anywhere at the moment.

`diverged` may not be needed when stuff is immutable, but I may just not really understand it yet! :sweat_smile: